src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Constructor.cs

```diff
@@ -0,0 +1,1365 @@
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics(
+            // (12,28): error CS1729: 'MyList<int>' does not contain a constructor that takes 0 arguments
+            //         MyList<int> list = [with(10)];
+            Diagnostic(ErrorCode.ERR_BadCtorArgCount, "[with(10)]").WithArguments("MyList<int>", "0").WithLocation(12, 28));
+    }
+
+    [Fact]
+    public void WithElement_ProtectedConstructor()
+    {
+        var source = """
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                protected MyList(int capacity) : base(capacity) { }
+            }
+            
+            class C
+            {
+                void M()
+                {
+                    MyList<int> list = [with(10)];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics(
+            // (12,28): error CS1729: 'MyList<int>' does not contain a constructor that takes 0 arguments
+            //         MyList<int> list = [with(10)];
+            Diagnostic(ErrorCode.ERR_BadCtorArgCount, "[with(10)]").WithArguments("MyList<int>", "0").WithLocation(12, 28));
+    }
+
+    [Fact]
+    public void WithElement_InternalConstructor_SameAssembly()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                internal MyList(int capacity) : base(capacity) 
+                {
+                    Console.Write("Internal constructor ");
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list = [with(10), 1];
+                    Console.WriteLine(list.Capacity + " " + list.Count);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("Internal constructor 10 1"));
+    }
+
+    #endregion
+
+    #region Constraint Tests
+
+    [Fact]
+    public void WithElement_TypeConstraints()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T> where T : class
+            {
+                public MyList(T item) : base()
+                {
+                    Add(item);
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<string> list = [with("first"), "second"];
+                    Console.WriteLine(list.Count);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("2"));
+    }
+
+    [Fact]
+    public void WithElement_TypeConstraints_Violation()
```

Comment:
Consider adding a test for `MyList<TElementType, TConstructorParameter>`, and at least verifying whether type inference can put it all together.

Reply:
c7c87747143